File: src/ObjectManager/ObjectManager/CellManager.cs
using OA.Core;
using System.Collections;
using UnityEngine;

namespace OA
{
    public interface ICellRecord : IRecord
    {
        bool IsInterior { get; }
        Color? AmbientLight { get; }
    }

    public class InRangeCellInfo
    {
        public GameObject GameObject;
        public GameObject ObjectsContainerGameObject;
        public ICellRecord CellRecord;
        public IEnumerator ObjectsCreationCoroutine;

        public InRangeCellInfo(GameObject gameObject, GameObject objectsContainerGameObject, ICellRecord cellRecord, IEnumerator objectsCreationCoroutine)
        {
            GameObject = gameObject;
            ObjectsContainerGameObject = objectsContainerGameObject;
            CellRecord = cellRecord;
            ObjectsCreationCoroutine = objectsCreationCoroutine;
        }
    }

    public class RefCellObjInfo
    {
        public object RefObj; //: CELLRecord.RefObjDataGroup
        public IRecord ReferencedRecord;
        public string ModelFilePath;
    }

    public interface ICellManager
    {
        Vector3Int GetCellId(Vector3 point, int worldId);
        InRangeCellInfo StartCreatingCell(Vector3Int cellId);
        InRangeCellInfo StartCreatingCellByName(int worldId, int cellId, string cellName);
        void UpdateCells(Vector3 currentPosition, int worldId, bool immediate = false, int cellRadiusOverride = -1);
        void DestroyAllCells();
    }
}

File: src.mac/ObjectManager/ObjectManager/ObjectManager/CellManager.swift
//
//  CellManager.swift
//  ObjectManager
//
//  Created by Sky Morey on 5/29/18.
//  Copyright © 2018 Sky Morey. All rights reserved.
//

import Foundation
import CoreGraphics
import SceneKit

public protocol ICellRecord : IRecord {
    var isInterior: Bool {get}
    //var ambientLight: CGColor? {get}
}

public class InRangeCellInfo {
    var gameObject: GameObject
    var objectsContainerGameObject: GameObject
    var cellRecord: ICellRecord
    var objectsCreationCoroutine: CoTask
    
    init(gameObject: GameObject, objectsContainerGameObject: GameObject, cellRecord: ICellRecord, objectsCreationCoroutine: CoTask) {
        self.gameObject = gameObject
        self.objectsContainerGameObject = objectsContainerGameObject
        self.cellRecord = cellRecord
        self.objectsCreationCoroutine = objectsCreationCoroutine
    }
}

public class RefCellObjInfo {
    var refObj: Any
    var referencedRecord: IRecord?
    var modelFilePath: String?
    
    init(refObj: Any, referencedRecord: IRecord? = nil, modelFilePath: String? = nil) {
        self.refObj = refObj
        self.referencedRecord = referencedRecord
        self.modelFilePath = modelFilePath
    }
}

public protocol ICellManager {
    func getCellId(point: float3, world: Int32) -> int3
    func startCreatingCell(cellId: int3) -> InRangeCellInfo?
    func startCreatingCellByName(world: Int32, cellId: Int, cellName: String) -> InRangeCellInfo?
    func updateCells(currentPosition: float3, world: Int32, immediate: Bool, cellRadiusOverride: Int32)
    func destroyAllCells()
}
